src/par_iter/slice.rs

```diff
@@ -26,40 +25,65 @@ impl<'data, T: Sync + 'data> IntoParallelRefIterator<'data> for [T] {
 
 impl<'data, T: Sync> ParallelIterator for SliceIter<'data, T> {
     type Item = &'data T;
-    type Shared = ();
-    type State = Self;
 
-    fn state(self) -> (Self::Shared, Self::State) {
-        ((), self)
+    fn drive_unindexed<'c, C: UnindexedConsumer<'c, Item=Self::Item>>(self,
+                                                                      consumer: C,
+                                                                      shared: &'c C::Shared)
+                                                                      -> C::Result {
+        bridge(self, consumer, &shared)
+    }
+}
+
+unsafe impl<'data, T: Sync> BoundedParallelIterator for SliceIter<'data, T> {
+    fn upper_bound(&mut self) -> usize {
+        ExactParallelIterator::len(self)
+    }
+
+    fn drive<'c, C: Consumer<'c, Item=Self::Item>>(self,
+                                                   consumer: C,
+                                                   shared: &'c C::Shared)
+                                                   -> C::Result {
+        bridge(self, consumer, &shared)
     }
 }
 
-unsafe impl<'data, T: Sync> BoundedParallelIterator for SliceIter<'data, T> { }
+unsafe impl<'data, T: Sync> ExactParallelIterator for SliceIter<'data, T> {
+    fn len(&mut self) -> usize {
+        self.slice.len() as usize
+    }
+}
 
-unsafe impl<'data, T: Sync> ExactParallelIterator for SliceIter<'data, T> { }
+impl<'data, T: Sync> IndexedParallelIterator for SliceIter<'data, T> {
```

Comment:
@cuviper 

> I need T: 'data for this to compile on stable, which makes sense since SliceProducer requires it. Odd that rust nightly doesn't care though...

Thanks. I think this is because of RFC 1214, which allows impls to infer that `T: 'data` because `SliceIter` required it (that code only became finalized in nightly).